File: watchers/osv.dart
import 'http.dart' as http;

typedef JsonObject = Map<String, Object?>;
typedef OsvPoster = Future<Object?> Function(JsonObject body);
typedef OsvQuery = Future<List<JsonObject>> Function(
  String ecosystem,
  String name,
);

final _advisoryId = RegExp(r'^[A-Za-z0-9][A-Za-z0-9_.:-]{0,127}$');
final _endpoint = Uri.parse('https://api.osv.dev/v1/query');

bool isSafeAdvisoryId(String value) => _advisoryId.hasMatch(value);

Future<List<JsonObject>> queryPackage(
  String ecosystem,
  String name, {
  OsvPoster? post,
}) async {
  final send = post ?? (body) => http.postJson(_endpoint, body);
  final records = <JsonObject>[];
  String? pageToken;
  while (true) {
    final query = <String, Object?>{
      'package': <String, String>{'name': name, 'ecosystem': ecosystem},
      if (pageToken != null) 'page_token': pageToken,
    };
    final payload = _object(await send(query), 'OSV response');
    final rawVulnerabilities = payload['vulns'] ?? const <Object?>[];
    if (rawVulnerabilities is! List) {
      throw const FormatException('OSV vulnerabilities was not a list');
    }
    for (final rawVulnerability in rawVulnerabilities) {
      final vulnerability = _object(
        rawVulnerability,
        'OSV vulnerability',
      );
      final id = vulnerability['id'];
      if (id is! String || !isSafeAdvisoryId(id)) {
        throw FormatException('OSV returned an unsafe advisory ID: $id');
      }
      records.add(vulnerability);
    }
    final rawPageToken = payload['next_page_token'];
    if (rawPageToken == null || rawPageToken == '') {
      return records;
    }
    if (rawPageToken is! String || rawPageToken.length > 4096) {
      throw const FormatException('OSV returned an invalid page token');
    }
    pageToken = rawPageToken;
  }
}

JsonObject _object(Object? value, String label) {
  if (value is! Map) {
    throw FormatException('$label was not an object');
  }
  final result = <String, Object?>{};
  for (final entry in value.entries) {
    if (entry.key is! String) {
      throw FormatException('$label had a non-string key');
    }
    result[entry.key as String] = entry.value;
  }
  return result;
}

File: watchers/osv.py
"""Small, fail-loud OSV client shared by repository watchers."""

from __future__ import annotations

import json
import re
import urllib.error
import urllib.request

_ADVISORY_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_.:-]{0,127}$")


def query_package(ecosystem: str, name: str) -> list[dict[str, object]]:
    """Return every OSV record for a package, following pagination."""

    records: list[dict[str, object]] = []
    page_token: str | None = None
    while True:
        query: dict[str, object] = {
            "package": {"name": name, "ecosystem": ecosystem}
        }
        if page_token:
            query["page_token"] = page_token
        request = urllib.request.Request(
            "https://api.osv.dev/v1/query",
            data=json.dumps(query).encode(),
            headers={
                "Content-Type": "application/json",
                "User-Agent": "keybay-security-watchers/1",
            },
        )
        with urllib.request.urlopen(request, timeout=30) as response:
            try:
                payload = json.load(response)
            except json.JSONDecodeError as error:
                raise urllib.error.URLError(
                    f"non-JSON OSV response: {error}"
                ) from error
        if not isinstance(payload, dict):
            raise urllib.error.URLError("OSV response was not an object")
        vulnerabilities = payload.get("vulns") or []
        if not isinstance(vulnerabilities, list):
            raise urllib.error.URLError("OSV vulnerabilities was not a list")
        for vulnerability in vulnerabilities:
            if not isinstance(vulnerability, dict):
                raise urllib.error.URLError("OSV vulnerability was not an object")
            advisory_id = vulnerability.get("id")
            if not isinstance(advisory_id, str) or not _ADVISORY_ID.fullmatch(
                advisory_id
            ):
                raise urllib.error.URLError(
                    f"OSV returned an unsafe advisory ID: {advisory_id!r}"
                )
            records.append(vulnerability)
        raw_page_token = payload.get("next_page_token")
        if raw_page_token is None or raw_page_token == "":
            return records
        if not isinstance(raw_page_token, str) or len(raw_page_token) > 4096:
            raise urllib.error.URLError("OSV returned an invalid page token")
        page_token = raw_page_token
